packages/sanes-chrome-extension/src/extension/bsMessageHandler/externalHandler.ts

```diff
@@ -0,0 +1,62 @@
+/*global chrome*/
+import { JsonRpcErrorResponse, JsonRpcRequest } from '@iov/jsonrpc';
+import { UseOnlyJsonRpcSigningServer } from '../../logic/persona';
+
+type SigningServer = UseOnlyJsonRpcSigningServer | undefined;
+type SendResponse = (response: any) => void; // eslint-disable-line @typescript-eslint/no-explicit-any
+
+function handleWithServer(
+  signingServer: SigningServer,
+  request: JsonRpcRequest,
+  sendResponse: SendResponse
+): void {
+  if (!signingServer) {
+    console.warn('Could not pass message to signing server');
+    const error: JsonRpcErrorResponse = {
+      jsonrpc: '2.0',
+      id: typeof request.id === 'number' ? request.id : null,
+      error: {
+        code: -32000,
+        message: 'Signing server not ready',
+      },
+    };
+    sendResponse(error);
+    return;
+  }
+
+  signingServer
+    .handleUnchecked(request)
+    .then(response => {
+      console.log('Responding', response);
+      sendResponse(response);
+    })
+    .catch(console.error);
+}
+
+function isForServer(request: JsonRpcRequest): boolean {
+  const { method } = request;
+  return method === 'getIdentities' || method === 'signAndPost';
+}
+
+function handleResponseToExtension(request: JsonRpcRequest, sendResponse: SendResponse): void {
+  chrome.browserAction.setIcon({ path: 'assets/icons/request128.png' });
+  chrome.browserAction.setBadgeBackgroundColor({ color: [0, 0, 0, 0] });
+  chrome.browserAction.setBadgeText({ text: 'REQ' });
+
+  // TODO dispath info to redux for getting the info in the extension. Include the
+  // sendResponse ref for handling answer
+  sendResponse(`I have handled the ${request.method} request changing the icon`);
+}
+
+export function handleExternalMessage(
+  signingServer: SigningServer,
+  request: JsonRpcRequest,
+  sendResponse: SendResponse
+): void {
+  if (isForServer(request)) {
+    handleWithServer(signingServer, request, sendResponse);
+    return;
+  }
+
+  handleResponseToExtension(request, sendResponse);
```

Comment:
I don't know if we really need requests that do no go to the signing server. I think everything should go there. However, let's discuss what this is for.

Reply:
This is a good question. I also thought that as well, when I saw Albert's ticket, I figured out introduce extra logic for handling user interaction, leaving everything to signing server would increase difficulty in the way I understand it. 

Let's think in the next feature: website foo.com asking for being granted to get identities:
```
OPTION A: Everything goes to signing server
1. The website sends a message to background script `getIdentities`
2. Background script blocks the flow generates logic showing a view in the extension if the website can be whitelisted for asking for identities
3. Wait for user interaction
4. Handle response
```

```
OPTION B: WE create a separate message
1. The website sends a message to background script `getIdentities`
2. Background script goes for other routine focused on user interaction
3. BAckground script creates a pool of pending requests
4. Extension handles those requests
5. Extension answers
6. Background script answers
```

Despite OPTION B increase a bit the logic in the background script, I prefer to implement it because:
- I do not modify the signing server logic
- We keep everything pure (one part only does one thing)
- I have a clear idea about how to handle the logic between the background script and the extension in a clean way
- Easier to test

Seeing it won't take too much time (I hope in less than one hour I am going to finish it), I will go with option b and once it is finished, let's talk again. What do you think?

Reply:
We need to have something like B anyway. The problem with A.2 is that the BS itself cannot show anything to the UI.

However, the question was another one: are all requests passed to the signing server (one way or another) or does the BS need to expose functionality that is not in the signing server.

Reply:
if you consider "keep a pool of requests" in the BS which requires user interaction as an exposed functionality, then yes.

I see the BS something like "Ey! I am the BS you can ask me for: "getIdentities | signAndPost | foo | bar | baz", the first two are handled directly in the BS because we have the signing server available, the other ones are enqueued because they need to be sorted with user interaction in the extension itself.

What we can do, and is something I really like is "force" to unify requests, despite "foo | bar | baz" are not handled in signing server, those requests should have the same type as getIdentities (for example) for consistency. 

Maybe I have answered your question better now?

Reply:
I would not offer "getIdentities | signAndPost | foo | bar | baz" as long as there is no "foo | bar | baz".

If the website by accident sends a method other than getIdentities | signAndPost, the signing server automatically responds with a MethodNotFoundError, so BS should pass everything to the signing server.

Reply:
Of course, we can have some checks for assuring the method received is an allowed one. In fact, this is a must.

The 'foo | bar | baz' are just examples, as we previously discussed, how to handle situations like what Albert is doing for "accept website to ask for identities". Again, in general, you can think of 'foo | bar | baz', as msgs for handling stuff on the extension which needs user interaction rather what the signing server does. 

Please, I do need agreement on this, for unblocking me. Reasons for separation concerns were listed before.